<a href="https://colab.research.google.com/github/torrancefredell/wrangling/blob/main/wrangling_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment: Data Wrangling


In [ ]:
! git clone https://github.com/DS3001/wrangling

Cloning into 'wrangling'...
remote: Enumerating objects: 63, done.
remote: Counting objects: 100% (24/24), done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 63 (delta 18), reused 7 (delta 7), pack-reused 39
Receiving objects: 100% (63/63), 6.25 MiB | 18.13 MiB/s, done.
Resolving deltas: 100% (24/24), done.


<font size="5">**Q1**</font> Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

<font size="5">**Q1 Responses**</font>
1. This paper is about a framework that can be used to "tidy" datasets by developing tools and streamlining processes for this purpose, so that datasets can be easily analyzed and visualized.
2. The tidy data standard was made to simplify data exploration and analysis, as well as to condense the number of tools used in order to spend more time drawing valuable insights from data, rather than trying to figure out how to fix it up.
3. The first quote, which compares happy vs. messy datasets and families, aims to share the message that all happy families and datasets are happy in one way; they do not have major issues, and as a result they are all similar, with a common structure. However, just as a family could appear messy because of any number of factors, datasets are the same; they could be messy because of too many missing pieces, faulty column/row naming methods, incorrect data types, etc. The second sentence refers to the complexities of understanding what is a variable versus observation in a dataset. This is difficult because variables and observations are very much defined in the context of the study that they are a part of, so it is difficult to say that they will universally "look" a certain way or describe certain things.
4. Values are data points, of which the dataset is mainly comprised. Variables are the various attributes of whatever is being studied, and observations are the different units that are measured for each variable.
5. Tidy data is defined as data that has three requirements: every variable and observation is its own column and row, respectively, and all of the observational units form the table itself.
6. The five most common problems with messy datasets are as follows: column headers are values when they should be variable names, columns contain multiple variables, variables can be found in both rows and columns, multiple types of observational units are in one table, and/or an observational unit can be found in multiple tables. The data in table four are messy because the column headers are values when they should be variable names. The variables are religion, income, and frequency, but the column headers (income ranges) should correspond with values in one column titled "income" inside the table. Melting a dataset refers to the process of taking several columns and combining them into a single column with some 'variable' header, as well as an associated 'value' column, which contains the original columns' data. Each original column header is thus a value in the new variable column.  
7. Dataset 11 is messy for a few reasons: first, it has an excessive number of blank spots, marked by "--." Second, each day of the month has its own column, when they could instead be condensed down to a single "dates" column with an associated "values" column next to it. Dataset 12 is tidy and molten because the 30+ columns dedicated to the date have been condensed to one column, which has an associated value column.
8. The "chicken and egg" problem refers to the problem where tidy data and tidy tools are interdependent; either one could come first, but realistically, neither one could be properly developed without the other. Therefore, they must be simultaneously adopted and advanced. In the future, Wickham hopes that scientists will build off of and improve the tidy data framework. Additionally, he hopes that tools that work with values stored in multidimensional arrays will be created, as well as tools that can choose between array-tidy and dataframe-tidy formats, for memory and performance sakes. Along with tidying, other tools that he mentions related to data cleaning include ones that could potentially find and missing values or typos, parse dates, work with international data, etc.

<font size="5">**Q2**</font> This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

<font size="5">**Q2. Part 1**</font>
- Clean Price variable
- Explain choices
- How many missing values?

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv('wrangling/assignment/data/airbnb_hw.csv', low_memory=False)
df.head(3)

Host Id Host Since                             Name Neighbourhood   \
0   5162530        NaN  1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN  Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN             Sunny Room in Harlem      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   

   Number of Records  Number Of Reviews Price  Review Scores Rating  
0                  1                  0   145                   NaN  
1                  1                  1    37                   NaN  
2                  1                  1    28                   NaN

In [ ]:
df.shape, '\n'

((30478, 13), '\n')

In [ ]:
price = df['Price']
price.unique()[:10]

array(['145', '37', '28', '199', '549', '149', '250', '90', '270', '290'],
      dtype=object)

In [ ]:
price = price.str.replace(',','')
price = pd.to_numeric(price,errors='coerce')
price

0        145
1         37
2         28
3        199
4        549
        ... 
30473    300
30474    125
30475     80
30476     35
30477     80
Name: Price, Length: 30478, dtype: int64

In [ ]:
missing_vals_count = price.isna().sum()
missing_vals_count

0

In [ ]:
df['Price'] = price
del price
df['Price'].tail(3)

30475    80
30476    35
30477    80
Name: Price, dtype: int64

<font size="4">**Explanation of choices made**</font>

First, I did a quick visual check of the Price column, and noticed that the data type was 'object' instead of int64. I made this adjustment for manipulation purposes. Then, I checked that there were no NaN values, which there weren't.

<font size="5">**Q2. Part 2**</font>
- Clean Type variable
- Explain choices

In [ ]:
sdf = pd.read_csv('wrangling/assignment/data/sharks.csv', low_memory=False)
sdf.head(3)

index   Case Number                  Date    Year        Type    Country  \
0      0    2020.02.05           05-Feb-2020  2020.0  Unprovoked        USA   
1      1  2020.01.30.R  Reported 30-Jan-2020  2020.0    Provoked    BAHAMAS   
2      2    2020.01.17           17-Jan-2020  2020.0  Unprovoked  AUSTRALIA   

              Area       Location                  Activity             Name  \
0             Maui            NaN  Stand-Up Paddle boarding              NaN   
1           Exumas            NaN                  Floating  Ana Bruna Avila   
2  New South Wales  Windang Beach                   Surfing   Will Schroeter   

   ... Unnamed: 246 Unnamed: 247 Unnamed: 248 Unnamed: 249 Unnamed: 250  \
0  ...          NaN          NaN          NaN          NaN          NaN   
1  ...          NaN          NaN          NaN          NaN          NaN   
2  ...          NaN          NaN          NaN          NaN          NaN   

  Unnamed: 251 Unnamed: 252 Unnamed: 253 Unnamed: 254 Unnamed: 255  
0          NaN          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN          NaN  

[3 rows x 257 columns]

In [ ]:
type = sdf['Type']
type.unique()

array(['Unprovoked', 'Provoked', 'Questionable', 'Watercraft',
       'Unconfirmed', 'Unverified', 'Invalid', 'Under investigation',
       'Boating', 'Sea Disaster', nan, 'Boat', 'Boatomg'], dtype=object)

In [ ]:
type.value_counts(), '\n'

(Unprovoked             4716
 Provoked                593
 Invalid                 552
 Sea Disaster            239
 Watercraft              142
 Boat                    109
 Boating                  92
 Questionable             10
 Unconfirmed               1
 Unverified                1
 Under investigation       1
 Boatomg                   1
 Name: Type, dtype: int64,
 '\n')

In [ ]:
type = type.replace(['Invalid', 'Questionable','Unconfirmed','Unverified','Under investigation'], 'Invalid')
type = type.replace(['Sea Disaster', 'Watercraft','Boat','Boating','Boatomg'], 'Sea Disaster')
type.value_counts() # Sanity check

Unprovoked      4716
Provoked         593
Sea Disaster     583
Invalid          565
Name: Type, dtype: int64

In [ ]:
df['Type'] = type
del type
df['Type'].value_counts() # Sanity check

Unprovoked      4716
Provoked         593
Sea Disaster     583
Invalid          565
Name: Type, dtype: int64

<font size="4">**Explanation of choices made**</font>


First, I did a quick visual check of the Type column, and took a count of each type of attack. "Invalid," "Uncomfired," "Unverified," etc. seemed pretty similar in meaning, so I combined them and named the category the same as the most populated one originally ("Invalid"). Also, several different types of attacks were related to boats, so I combined all of those (including the "Boatomg" type, which was a typo). I left this category as "Sea Disasters."

<font size="5">**Q2. Part 3**</font>

- Clean WhetherDefendantWasReleasedPretrial variable
- Replace missing values with np.nan
- Explain choices

In [ ]:
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
wdf = pd.read_csv(url,low_memory=False)

In [ ]:
wdf.shape, '\n'

((22986, 709), '\n')

In [ ]:
released = wdf['WhetherDefendantWasReleasedPretrial']
released.unique()

array([9, 0, 1])

In [ ]:
released.value_counts()

1    19154
0     3801
9       31
Name: WhetherDefendantWasReleasedPretrial, dtype: int64

In [ ]:
released = released.replace(9,np.nan)
released.unique() # Sanity check

array([nan,  0.,  1.])

In [ ]:
wdf['WhetherDefendantWasReleasedPretrial'] = released # Replace data column with cleaned values
del released

<font size="4">**Explanation of choices made**</font>

First, I loaded the dataset from the website because it was big, according to the notes (confirmed with wdf.shape). Then, I checked what values were in the specified column, and how many there were. According to the the codebook I found that the 9's meant "unclear," so I changed those to np.nan, and then did a sanity check with .unique() to make sure it worked.

<font size="5">**Q2. Part 4**</font>

- Clean ImposedSentenceAllChargeInContactEvent variable
- Explain choices
- Hint: Look at SentenceTypeAllChargesAtConvictionInContactEvent variable

In [ ]:
imposed_sentence = wdf['ImposedSentenceAllChargeInContactEvent']
imposed_sentence = pd.to_numeric(imposed_sentence, errors='coerce')
sentence_type = wdf['SentenceTypeAllChargesAtConvictionInContactEvent']
print(imposed_sentence.dtype, sentence_type.dtype)

float64 int64


In [ ]:
print(imposed_sentence.unique()[:10], '\n', sentence_type.unique())

[         nan  60.          12.           0.98562628  36.
   6.          24.           5.9137577  120.          72.        ] 
 [9 0 1 4 2]


In [ ]:
missing_imposed_count = (imposed_sentence.isna()).sum()
missing_imposed_count

9053

In [ ]:
wdf['imposed_NA'] = imposed_sentence.isna()
print(pd.crosstab(wdf['imposed_NA'],sentence_type),'\n')

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
imposed_NA                                                                  
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274 



In [ ]:
imposed_sentence = imposed_sentence.mask(sentence_type == 4, 0)
imposed_sentence = imposed_sentence.mask(sentence_type == 9, np.nan)

In [ ]:
imposed_NA = imposed_sentence.isnull()
print(pd.crosstab(imposed_NA, sentence_type), '\n')

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914  8779    0
True                                                 0     0    0     0  274 



In [ ]:
wdf['ImposedSentenceAllChargeInContactEvent'] = imposed_sentence
del imposed_sentence, sentence_type

<font size="4">**Explanation of choices made**</font>

When I looked at imposed_sentence, some of the values had very long trailing decimals, and they were all strings. 9,053 of them were also missing. I made a dummy variable for the null imposed sentence values and used a crosstab function, which told me that a lot of imposed sentences for sentence type 4 were NaN. I used the codebook to see that these were cases that were dismissed / deferred, etc., so I changed their values to 0s, since these cases were similar to "no incarceration." I tried to use .loc to do this, but it was taking too long, so I used the mask() function instead. I did something similar for sentence type 9, which also had a lot of NaN imposed sentence values, and changed those values to np.nan.

<font size="5">**Q3**</font> Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

<font size="5">**Q3 Responses**</font>
1. According to the US Census Bureau, race data is collected via self-identification, and individuals may choose to report more than one race group. They could check a box for things like "White" or "African Am" and then specify origin ("Irish" or "Jamaican," for example) through print on the paper 2020 census.
2. Race data are collected in order to aid governments and communities
in their fight against discrimination. When used in conjunction with data related to housing, education, employment, etc., race data become a powerful tool to uncover inequalities that might otherwise be difficult to detect. These data collectively also allow for more just policy-making, since they help politicians understand demographic trends and the distribution of various resources. Good quality data is needed so that this entire process is accurate and so that policies can reflect the needs of the populations that they affect.
3. I think that the Census did a good job of trying to make sure that people from all racial backgrounds could be accounted for. They not only offered opportunities for individuals to state their race, but also their specific origin. Additionally, people could choose more than one race, which was a positive step toward accomodating mixed-race people. One specific issue that I noticed was that the space for American Indian or Alaskan Native may be insufficient; it does not provide enough space even for the origin examples included on the Census. People belonging to these groups who see this may be confused or frustrated with this oversight, which could potentially lead to underreporting from those populations.
4. The same inclusivity that was seen in the race section of the Census generally extended to queer individuals as well. However, I think there's room for improvement in how the Census addresses 'sex.' Given the evolving concept of gender diversity, the options of 'male' or 'female' may be too restrictive for some. To this point, incorporating a broader range of gender identities or perhaps adding a print-in section may be beneficial for the sake of political correctness. The makers of similar surveys could follow suit and adopt more open-ended, inclusive questions and options.
5. When it comes to cleaning data, I am concerned that the aforementioned characteristics might be simplified into their overarching descriptors. For example, although someone writes in their specific origin, for the sake of data manipulation, scientists might just overgeneralize and say they are just a part of the broader category, and lose all distinction between various backgrounds. People who are not trained properly may introduce bias into the data when cleaning it. Additionally, missing values may cause problems because they would make it harder for data analysts to find patterns among and between connected data, and the statistical power of the dataset might be impacted. Good practices might include not adjusting certain data, such as that involved with race or sex, since the individuals who filled out the Census probably did it a certain way on purpose. These decisions can be documented so that analysts later on can understand why they were made. If imputation is necessary, it should be done in such a way that reflects the uncertainty in the dataset. Along with what was mentioned before, bad practice would include filling in data or throwing away essential pieces of data (overgeneralizing).
6. If someone were to develop such an algorithm, I would have a few concerns: primarily, the imputed values might not reflect the affected individuals correctly, leading to misrepresentation. In turn, this could have consequences when it comes to policy making, as mentioned before. Misrepresentation can also reinforce stereotypes and marginalize certain minority groups. Additionally, I would worry that the algorithm would not adequately consider the dynamics of nuanced matters such as race or sex over time, which are influenced by personal experiences, cultural shifts, and/or social attitudes.